### PR TITLE
fix: profile payload issues

### PR DESCRIPTION
**Pull request by armcknight**
There were a couple remaining issues with profile payloads:
- sometimes we still try to send an envelope with 0 samples, presumably for very fast start/stop of profiler. we simply discard these and log for now
- we weren't deduplicating frames across an entire profile, only per sample
- we weren't performing a correct deep equality check between stacks

### CODE REVIEW

File: Sources/Sentry/SentryProfiler.mm [hunk 1]
```diff
@@ -64,6 +64,99 @@
     return [symbolNSStr substringWithRange:[match rangeAtIndex:1]];
 }
 
+void
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:67]: This diff is hard to parse since it was extracted out into a new function, what are the changes?

File: Sources/Sentry/SentryProfiler.mm [hunk 2]
```diff
@@ -622,6 +649,11 @@ - (void)captureEnvelope
     }
     profile[@"transactions"] = transactionsInfo;
 
+    if ([((NSArray *)profile[@"profile"][@"samples"]) count] == 0) {
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:652]: Couldn't we just check `[samples count]`? - it's a local var in this scope I believe

Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:652]: ooh good find

File: Sources/Sentry/SentryProfiler.mm [hunk 3]
```diff
@@ -435,12 +440,30 @@ - (void)start
...
+                            if (![nextStackFrame isEqualToNumber:stack[nextStackFrameIdx]]) {
+                                *stopFrameEnumeration = YES;
+                                found = NO;
+                            }
+                        }];
+                        if (found) {
+                            *stopStackEnumeration = YES;
+                        }
+                        return found;
+                    }];
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm]: This is an O(n^2) loop where n can be quite big so I would expect this to have some impact on overhead. Does this function get called on the main thread and do we know how long it takes for a larger (e.g. 10 seconds) profile?

Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm]: I converted this to a fast lookup using another NSDictionary. Thanks for catching this!

File: Sources/Sentry/SentryProfiler.mm [hunk 4]
```diff
@@ -64,6 +64,85 @@
...
 
+void
+processBacktrace(const Backtrace &backtrace,
+    NSMutableDictionary<NSString *, NSMutableDictionary *> *threadMetadata,
+    NSMutableDictionary<NSString *, NSDictionary *> *queueMetadata,
+    NSMutableArray<NSDictionary<NSString *, id> *> *samples,
+    NSMutableArray<NSMutableArray<NSNumber *> *> *stacks,
+    NSMutableArray<NSDictionary<NSString *, id> *> *frames,
+    NSMutableDictionary<NSString *, NSNumber *> *frameIndexLookup, uint64_t startTimestamp,
+    NSMutableDictionary<NSArray<NSNumber *> *, NSNumber *> *stackIndexLookup)
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:75]: do we know if using an `NSArray` as a dictionary key works as expected? I'm guessing it computes a reasonable hash but just wondering

**Review by indragiek — APPROVED** [review]

File: Sources/Sentry/SentryProfiler.mm (hunk 4 above)
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:75]: We're using an `NSArray<NSNumber *>`, and both of those classes conform to NSCoding and NSCopying which are the requirements for NSDictionary keys. The tests that passed with the slow comparison still pass with the hash approach.

But now that you mention it, copying the whole array to use as a key is going to be slower and less memory efficient than just computing the hash and using that as the key. I will make that change.

Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm]: OK, wound up going the route of constructing string keys from the stacks since NSArray.hash only returns the array length ಠ_ಠ
